src/master/heartbeat_cluster.cpp

```diff
@@ -383,40 +372,45 @@ namespace cubhb
     return NULL;
   }
 
-  void
-  cluster::remove_ui_node (ui_node *&node)
+  node_entry *
+  cluster::find_node_except_me (const cubbase::hostname_type &node_hostname) const
   {
-    if (node == NULL)
+    for (node_entry *node : nodes)
       {
-	return;
+	if (node->get_hostname () != node_hostname || m_hostname == node_hostname)
+	  {
+	    continue;
+	  }
+
+	return node;
       }
 
-    ui_nodes.remove (node);
-    delete node;
-    node = NULL;
+    return NULL;
   }
 
   void
   cluster::cleanup_ui_nodes ()
   {
     std::chrono::system_clock::time_point now = std::chrono::system_clock::now ();
-    for (ui_node *node : ui_nodes)
+    for (std::list<ui_node *>::iterator it = ui_nodes.begin (); it != ui_nodes.end (); ++it)
       {
-	if ((now - node->last_recv_time) > UI_NODE_CLEANUP_TIME_IN_MSECS)
+	if ((now - (*it)->last_recv_time) > UI_NODE_CLEANUP_TIME_IN_MSECS)
 	  {
-	    remove_ui_node (node);
+	    auto tmp = it;
```

Comment:
1. What is auto type here? Note: auto should be used when type is self-explanatory for developers too :).
2. delete *tmp indicates that iterator was not dereferenced. Is it safe?

Reply:
fixed